### ISSUE TITLE
error saving files

### ISSUE THREAD

**Issue by matinashtiani** [sec-edgar/sec-edgar#180]
I installed the API and run the next couple of code lines. 
I got an error when I wanted to save the filing.

```python
from secedgar.filings import FilingType, Filing
from datetime import datetime

filing = Filing(cik_lookup='aapl',
                filing_type=FilingType.FILING_10K,
                start_date=datetime(2015, 1, 1),
                end_date=datetime(2019, 1, 1))

filing.save('./reports')
```

The following error is rising when I run the last command:

`---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-9-ae9d2e927b58> in <module>
----> 1 filing.save('FFS')

~\.conda\envs\Keras\lib\site-packages\secedgar\filings\filing.py in save(self, directory, dir_pattern, file_pattern)
    222 
    223         loop = asyncio.get_event_loop()
--> 224         loop.run_until_complete(self.client.wait_for_download_async(inputs))

~\.conda\envs\Keras\lib\asyncio\base_events.py in run_until_complete(self, future)
    568         future.add_done_callback(_run_until_complete_cb)
    569         try:
--> 570             self.run_forever()
    571         except:
    572             if new_task and future.done() and not future.cancelled():

~\.conda\envs\Keras\lib\asyncio\base_events.py in run_forever(self)
    523         self._check_closed()
    524         if self.is_running():
--> 525             raise RuntimeError('This event loop is already running')
    526         if events._get_running_loop() is not None:
    527             raise RuntimeError(

RuntimeError: This event loop is already running
10it [00:01,  9.88it/s]`

**Comment by jackmoody11** [comment]
Hi @matinashtiani, thanks for reporting this. Which version of secedgar are you using?

**Comment by fyx99** [comment]
Quick workaround -> probably problem for all using a newer python version with asyncio
!pip install nest_asyncio
import nest_asyncio
nest_asyncio.apply()

**Comment by jackmoody11** [comment]
> Quick workaround -> probably problem for all using a newer python version with asyncio
> 
> !pip install nest_asyncio
> 
> import nest_asyncio
> 
> nest_asyncio.apply()

@fyx99 is this a 3.9 issue? Or does it begin earlier than that?

**Comment by fyx99** [comment]
> > Quick workaround -> probably problem for all using a newer python version with asyncio
> > !pip install nest_asyncio
> > import nest_asyncio
> > nest_asyncio.apply()
> 
> @fyx99 is this a 3.9 issue? Or does it begin earlier than that?

probapbly, I am running on 3.9

**Comment by reteps** [comment]
@jackmoody11 do the tests run on 3.9?

**Comment by jackmoody11** [comment]
Yes, they do @reteps. Not sure why they aren't catching this.

**Comment by matinashtiani** [comment]
> Hi @matinashtiani, thanks for reporting this. Which version of secedgar are you using?

I am using 0.3.2.

**Comment by jackmoody11** [comment]
> > Hi @matinashtiani, thanks for reporting this. Which version of secedgar are you using?
> 
> 
> 
> I am using 0.3.2.
> 
> Also, my python version is 3.7.6

Ok, thanks for sharing this @matinashtiani. I will see if I can find the issue.

**Comment by jackmoody11** [comment]
@reteps I think this might have something to do with [this](https://stackoverflow.com/a/56662635/8206432).

**Comment by reteps** [comment]
I think we just add nest_asyncio as a dependency.

**Comment by jackmoody11** [comment]
> I think we just add nest_asyncio as a dependency.

@reteps Would rather find the underlying issue and fix if we aren't using asyncio like it is intended/supposed to be used.

**Comment by esvhd** [comment]
Hi - I ran into this problem with python 3.8.5 as well on Ubuntu. So far, I'm using `nest_asyncio` to get around this problem. Looking forward to a proper fix. Thanks.

**Comment by esvhd** [comment]
Actually correcting myself... using `nest_asyncio` did not solve the problem, it appeared to be doing something, but no file was downloaded into the specified folder...

**Comment by reteps** [comment]
@jackmoody11 [This answer](https://stackoverflow.com/a/56662635/15261182) gives us a explanation why this is illegal:


```
run_until_complete is not for running any number of arbitrary async functions synchronously, it is for running the main entry point of your entire async program. This constraint is not immediately apparent from the docs.

Since libraries like aiohttp will queue it's own entry point to run as a server and block the loop's synchronous operations using run_until_complete or run_forever, the event loop will already be running and you won't be able to run independent synchronous operations on that event loop and wait for it's result within that thread.
```

**Comment by reteps** [comment]
I'm not entirely sure how to resolve it, but the issue is most certainly in the [`wait_for_download_async` function](https://github.com/sec-edgar/sec-edgar/blob/bb0ea3beaa4d7c527a14ac81bf7987cf56de54d1/secedgar/client/network_client.py#L182)

**Comment by reteps** [comment]
Note:         `async with await session.get(link) as response:` -> `async with session.get(link) as response:` since we batch, we don't need to wait for session.

**Comment by reteps** [comment]
Final notes:

+ move client and conn to init of network client
+ remove fetch method, move into fetch_and_save method
+ make fetch_and_save global method, remove session parameter

I'll open PR to fix these, though I'm not sure that will resolve the issue.

**Comment by reteps** [comment]
Ok, it is added to 0.4.0

**Comment by esvhd** [comment]
Hi @reteps , thanks for this, has this code been merged to master? If so I'm happy to test it.

**Comment by reteps** [comment]
It isn't because the code works fine without it on my machine haha. Just ran your script:

```python3
retep@desktop:~/repos/secedgar_test$ python3 --version
Python 3.8.5
retep@desktop:~/repos/secedgar_test$ python3
Python 3.8.5 (default, Jul 28 2020, 12:59:40)
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import secedgar
>>> secedgar.__version__
'0.3.2'
>>>
retep@desktop:~/repos/secedgar_test$ python3 main.py
10it [00:01,  5.11it/s]
```

**Comment by reteps** [comment]
@esvhd Can you share your version of `aiohttp`?

```
>>> import aiohttp
>>> aiohttp.__version__
'3.6.3'
```

**Comment by esvhd** [comment]
Here's mine. Got installed when I pip installed `secedgar`

```
# Name                    Version                   Build  Channel
aiohttp                   3.7.4                    pypi_0    pypi
```

I changed it to your version 3.6.3 but it still did not work.. Here's what I need:


```
# had to use nest_asyncio, otherwise code fails...
import nest_asyncio
nest_asyncio.apply()

from secedgar.filings import Filing, FilingType

# 10Q filings for Apple (ticker "aapl")
my_filings = Filing(cik_lookup='aapl', filing_type=FilingType.FILING_10Q)

my_filings.save('~/data/edgar')
# this runs with tqdm status bar.... but...
```

Then when in the folder, there is not data...

**Comment by reteps** [comment]
I don't understand how mine could work but yours doesn't

**Comment by reteps** [comment]
What platform are you on? I am running this on `ubuntu 18.04` inside windows WSL2.

**Comment by reteps** [comment]
I just upgraded to the latest version of aiohttp which did not change anything, so I suspect it has to do with the python version if that is the only difference.

**Comment by reteps** [comment]
Possible problem: 3.8 has a different default event loop.

https://www.python.org/downloads/release/python-381/

Python3.8: `on Windows, the default asyncio event loop is now ProactorEventLoop`

**Comment by reteps** [comment]
Can you upgrade python to any version >=3.8.1 and test?

**Comment by esvhd** [comment]
I’m running Ubuntu 20.04..

**Comment by esvhd** [comment]
Hi - I’m actually running Pythian 3.8.5 here.. odd

**Comment by reteps** [comment]
So to confirm, we have identical python, sec-edgar, and platform, but mine works and yours does not.

**Comment by jackmoody11** [comment]
@esvhd Can you show us what you get when you run `pip list`?

**Comment by esvhd** [comment]
Appreciate your help folks. 

Here's my python details:

```
Python implementation: CPython
Python version       : 3.8.5
IPython version      : 7.19.0

nest_asyncio: 1.5.1
aiohttp     : 3.7.4
secedgar    : 0.3.2

Compiler    : GCC 7.3.0
OS          : Linux
Release     : 5.8.0-44-generic
Machine     : x86_64
Processor   : x86_64
CPU cores   : 20
Architecture: 64bit
```


Here's pip output:

```
Package                            Version             Location
---------------------------------- ------------------- ----------------------------
absl-py                            0.11.0
aiohttp                            3.7.4
alabaster                          0.7.12
anaconda-client                    1.7.2
anaconda-navigator                 1.10.0
anaconda-project                   0.9.1
appdirs                            1.4.4
arch                               4.2+4.g3d3b390
argh                               0.26.2
argon2-cffi                        20.1.0
asn1crypto                         1.4.0
astroid                            2.5
astropy                            4.2
astunparse                         1.6.3
async-generator                    1.10
async-timeout                      3.0.1
atomicwrites                       1.4.0
attrs                              20.3.0
autopep8                           1.5.4
Babel                              2.9.0
backcall                           0.2.0
backports.functools-lru-cache      1.6.1
backports.shutil-get-terminal-size 1.0.0
backports.tempfile                 1.0
backports.weakref                  1.0.post1
beautifulsoup4                     4.9.3
bitarray                           1.6.3
bkcharts                           0.2
black                              19.10b0
bleach                             3.3.0
blpapi                             3.14.0
bokeh                              2.2.3
boto                               2.49.0
Bottleneck                         1.3.2
brotlipy                           0.7.0
cachetools                         4.2.1
certifi                            2020.12.5
cffi                               1.14.5
chardet                            3.0.4
click                              7.1.2
cloudpickle                        1.6.0
clyent                             1.2.2
colorama                           0.4.4
colorcet                           2.0.6
conda                              4.9.2
conda-build                        3.20.5
conda-package-handling             1.7.2
conda-verify                       3.4.2
contextlib2                        0.6.0.post1
cryptography                       3.3.1
cvxopt                             1.2.0
cvxpy                              1.1.7
cycler                             0.10.0
Cython                             0.29.22
cytoolz                            0.11.0
dask                               2020.12.0
decorator                          4.4.2
defusedxml                         0.6.0
diff-match-patch                   20200713
distributed                        2021.2.0
docutils                           0.16
ecos                               2.0.7.post1
entrypoints                        0.3
et-xmlfile                         1.0.1
fastcache                          1.1.0
filelock                           3.0.12
filterpy                           1.4.1
flake8                             3.8.4
Flask                              1.1.2
flatbuffers                        1.12
fsspec                             0.8.3
future                             0.18.2
gast                               0.3.3
gevent                             21.1.1
glob2                              0.7
gmpy2                              2.0.8
google-auth                        1.27.0
google-auth-oauthlib               0.4.2
google-pasta                       0.2.0
greenlet                           1.0.0
grpcio                             1.32.0
h5py                               2.10.0
hdbscan                            0.8.26
HeapDict                           1.0.1
holoviews                          1.14.0
html5lib                           1.1
idna                               2.10
imageio                            2.9.0
imagesize                          1.2.0
importlib-metadata                 2.0.0
iniconfig                          1.1.1
intervaltree                       3.1.0
ipdb                               0.13.4
ipykernel                          5.3.4
ipython                            7.19.0
ipython-genutils                   0.2.0
ipywidgets                         7.5.1
isort                              5.7.0
itsdangerous                       1.1.0
jdcal                              1.4.1
jedi                               0.17.2
jeepney                            0.6.0
Jinja2                             2.11.3
joblib                             1.0.0
json5                              0.9.5
jsonschema                         3.2.0
jupyter                            1.0.0
jupyter-client                     6.1.7
jupyter-console                    6.2.0
jupyter-contrib-core               0.3.3
jupyter-contrib-nbextensions       0.5.1
jupyter-core                       4.7.1
jupyter-highlight-selected-word    0.2.0
jupyter-latex-envs                 1.4.6
jupyter-nbextensions-configurator  0.4.1
jupyterlab                         2.2.6
jupyterlab-pygments                0.1.2
jupyterlab-server                  1.2.0
Keras-Preprocessing                1.1.2
keyring                            22.2.0
kiwisolver                         1.3.1
lazy-object-proxy                  1.5.2
libarchive-c                       2.9
lightgbm                           3.0.0.99
llvmlite                           0.34.0
locket                             0.2.1
lxml                               4.6.2
Markdown                           3.3.3
MarkupSafe                         1.1.1
matplotlib                         3.3.2
mccabe                             0.6.1
mistune                            0.8.4
mkl-fft                            1.3.0
mkl-random                         1.1.1
mkl-service                        2.3.0
mlfinlab                           0.12.0              
mliv                               0.0.1               
mock                               4.0.3
more-itertools                     8.7.0
mpmath                             1.1.0
msgpack                            1.0.2
multidict                          5.1.0
multipledispatch                   0.6.0
multitasking                       0.0.9
mypy-extensions                    0.4.3
natsort                            7.1.1
navigator-updater                  0.2.1
nbclient                           0.5.1
nbconvert                          6.0.7
nbformat                           5.0.8
nest-asyncio                       1.5.1
networkx                           2.5
nltk                               3.5
nose                               1.3.7
notebook                           6.1.5
numba                              0.51.2
numexpr                            2.7.2
numpy                              1.19.2
numpydoc                           1.1.0
oauthlib                           3.1.0
olefile                            0.46
openpyxl                           3.0.6
opt-einsum                         3.3.0
osqp                               0.6.1
packaging                          20.9
pandas                             1.2.0
pandas-flavor                      0.2.0
pandocfilters                      1.4.3
panel                              0.10.3
param                              1.10.1
parso                              0.7.0
partd                              1.1.0
path                               15.1.2
pathlib2                           2.3.5
pathspec                           0.7.0
pathtools                          0.1.2
patsy                              0.5.1
pep8                               1.7.1
pexpect                            4.8.0
pickleshare                        0.7.5
Pillow                             8.1.0
pip                                20.3.3
pkginfo                            1.7.0
pluggy                             0.13.1
ply                                3.11
pretty-errors                      1.2.19
prometheus-client                  0.9.0
prompt-toolkit                     3.0.8
protobuf                           3.15.2
psutil                             5.8.0
psycopg2                           2.8.6
ptyprocess                         0.7.0
py                                 1.10.0
pyasn1                             0.4.8
pyasn1-modules                     0.2.8
pycodestyle                        2.6.0
pycosat                            0.6.3
pycparser                          2.20
pyct                               0.4.8
pycurl                             7.43.0.6
pydocstyle                         5.1.1
pyerfa                             1.7.2
pyflakes                           2.2.0
Pygments                           2.8.0
pyjanitor                          0.20.10
pylint                             2.7.1
pyls-black                         0.4.6
pyls-spyder                        0.3.2
pyodbc                             4.0.0-unsupported
pyOpenSSL                          20.0.1
pyparsing                          2.4.7
pyportfolioopt                     1.2.7
pyrsistent                         0.17.3
PySocks                            1.7.1
pytest                             0.0.0
python-dateutil                    2.8.1
python-edgar                       3.0.2               
python-jsonrpc-server              0.4.0
python-language-server             0.36.2
pytorch-lightning                  0.4.9
pytz                               2021.1
pyviz-comms                        2.0.1
PyWavelets                         1.1.1
pyxdg                              0.27
PyYAML                             5.4.1
pyzmq                              20.0.0
QDarkStyle                         2.8.1
QtAwesome                          1.0.1
qtconsole                          5.0.2
QtPy                               1.9.0
regex                              2020.11.13
requests                           2.25.1
requests-oauthlib                  1.3.0
rise                               5.6.1
rope                               0.18.0
rsa                                4.7.2
Rtree                              0.9.4
ruamel-yaml                        0.15.87
sacremoses                         0.0.43
scann                              1.2.1
scikit-image                       0.17.2
scikit-learn                       0.24.1
scipy                              1.6.1
scs                                2.1.2
seaborn                            0.11.1
sec-edgar-downloader               4.0.2
secedgar                           0.3.2
SecretStorage                      3.3.1
Send2Trash                         1.5.0
setuptools                         51.0.0.post20201207
shap                               0.37.0
simplegeneric                      0.8.1
singledispatch                     0.0.0
sip                                4.19.13
six                                1.15.0
slicer                             0.0.7
snowballstemmer                    2.1.0
sortedcollections                  2.1.0
sortedcontainers                   2.3.0
soupsieve                          2.2
Sphinx                             3.5.1
sphinxcontrib-applehelp            1.0.2
sphinxcontrib-devhelp              1.0.2
sphinxcontrib-htmlhelp             1.0.3
sphinxcontrib-jsmath               1.0.1
sphinxcontrib-qthelp               1.0.3
sphinxcontrib-serializinghtml      1.1.4
sphinxcontrib-websupport           1.2.4
spyder                             4.2.1
spyder-kernels                     1.10.2
SQLAlchemy                         1.3.23
statsmodels                        0.12.2
sympy                              1.7.1
tables                             3.6.1
tabulate                           0.8.7
tblib                              1.7.0
tensorboard                        2.4.1
tensorboard-plugin-wit             1.8.0
tensorflow                         2.4.1
tensorflow-estimator               2.4.0
termcolor                          1.1.0
terminado                          0.9.2
testpath                           0.4.4
textdistance                       4.2.1
threadpoolctl                      2.1.0
three-merge                        0.1.1
tifffile                           2020.10.1
tokenizers                         0.9.4
toml                               0.10.1
toolz                              0.11.1
torch                              1.7.1
torchaudio                         0.7.0a0+a853dff
torchtext                          0.8.0a0+0f911ec
torchvision                        0.8.2
tornado                            6.1
tqdm                               4.54.1
traitlets                          5.0.5
transformers                       4.3.0.dev0          
typed-ast                          1.4.2
typing-extensions                  3.7.4.3
ujson                              4.0.2
unicodecsv                         0.14.1
urllib3                            1.26.2
watchdog                           1.0.2
watermark                          2.1.0
wcwidth                            0.2.5
webencodings                       0.5.1
Werkzeug                           1.0.1
wheel                              0.36.2
widgetsnbextension                 3.5.1
wrapt                              1.12.1
wurlitzer                          2.0.1
xarray                             0.16.2
xlrd                               2.0.1
XlsxWriter                         1.3.7
xlwt                               1.3.0
xmltodict                          0.12.0
yapf                               0.30.0
yarl                               1.6.3
yfinance                           0.1.55
zict                               2.0.0
zipp                               3.4.0
zope.event                         4.5.0
zope.interface                     5.2.0
```

**Comment by reteps** [comment]
After upgrading my stuff, and comparing:

Your stuff:
```
secedgar 0.3.2
beautifulsoup4 4.9.3
aiohttp 3.7.4
async-timeout 3.0.1
requests 2.25.1
urllib3 1.26.2
lxml 4.6.2
tqdm 4.54.1
```

My stuff:
```
secedgar                      0.3.2
beautifulsoup4                4.9.3
aiohttp                       3.7.4
async-timeout                 3.0.1
requests                      2.25.1
urllib3                       1.26.3
lxml                          4.6.2
tqdm                          4.54.1

3.8.5 (default, Jul 28 2020, 12:59:40)
[GCC 9.3.0]
```

Not sure how you got the processor information, but this is so incredibly strange!

**Comment by reteps** [comment]
```
>>> import platform
>>> platform.uname()
uname_result(system='Linux', node='desktop', release='4.19.128-microsoft-standard', version='#1 SMP Tue Jun 23 12:58:10 UTC 2020', machine='x86_64', processor='x86_64')
>>> platform.node()
'desktop'
>>> platform.release()
'4.19.128-microsoft-standard'
>>> platform.version()
'#1 SMP Tue Jun 23 12:58:10 UTC 2020'
>>> platform.machine()
'x86_64'
>>> platform.processor()
'x86_64'
>>> platform.architecture()
('64bit', 'ELF')
>>> platform.python_build()
('default', 'Jul 28 2020 12:59:40')
>>> platform.python_version_tuple()
('3', '8', '5')
>>> platform.python_compiler()
'GCC 9.3.0'
>>> platform.python_build()
('default', 'Jul 28 2020 12:59:40')
```

**Comment by esvhd** [comment]
Try the watermark package, pretty easy to get system info.

**Comment by jackmoody11** [comment]
@esvhd can you set up a new virtual environment where you only `pip install secedgar` and then run the script? And then let us know if the problem persists.

**Comment by reteps** [comment]
@esvhd Here is a link to a guide for something that @jackmoody11 outlined. https://help.dreamhost.com/hc/en-us/articles/115000695551-Installing-and-using-virtualenv-with-Python-3

**Comment by esvhd** [comment]
Hi guys, let me try this later and will let you know. Thanks!

**Comment by reteps** [comment]
@esvhd Just wanted to give you a reminder we are waiting on that.

**Comment by jackmoody11** [comment]
@esvhd I just merged a PR into master that adds tqdm as a requirement, so that should be fixed now. Working on finalizing v0.3.3 release so that we can release that.

**Comment by reteps** [comment]
> @esvhd I just merged a PR into master that adds tqdm as a requirement, so that should be fixed now. Working on finalizing v0.3.3 release so that we can release that.
> 
> Can you try running the code _without_ `nest_asyncio`? We are trying to determine if a fresh install will let you run the code without any issues or if it is a bug.

I think he did. I am going to try setting up a venv as well, see if anything changes, maybe it stops working.

**Comment by reteps** [comment]
Wait a second... I think there is a bug with `~` saves... can you try saving to `./files` or something.

**Comment by esvhd** [comment]
Looks like that's it! I tried `filing.save('./edgar')` and it worked. Guess it didn't recognise `~` as home directory? But somehow the process did not fail. 

Thanks v much for looking into this.

**Comment by esvhd** [comment]
> @esvhd I just merged a PR into master that adds tqdm as a requirement, so that should be fixed now. Working on finalizing v0.3.3 release so that we can release that.
> 
> Can you try running the code _without_ `nest_asyncio`? We are trying to determine if a fresh install will let you run the code without any issues or if it is a bug.

Hi @jackmoody11 - I just tried WITHOUT importing `nest_asyncio` in my newly created environment, after saving to `./edgar` as @reteps suggested, the download **worked as expected**.

FYI - Interestingly, in my old base conda environment, without `nest_asyncio`, with `./edgar` it still did not work.. 

Looks like it was related to `save()`.

**Comment by jackmoody11** [comment]
Glad we got that figured out! I think that we may have used to have some logic that would expand the user directory. Should definitely mention that in the documentation.

**Comment by reteps** [comment]
I am going to mark this as closed and open a new issue for expansion of `~`.
